transport/src/main/java/io/camunda/zeebe/transport/stream/impl/RemoteStreamImpl.java

```diff
@@ -8,27 +8,78 @@
 package io.camunda.zeebe.transport.stream.impl;
 
 import io.camunda.zeebe.transport.stream.api.RemoteStream;
+import io.camunda.zeebe.transport.stream.impl.ImmutableStreamRegistry.AggregatedRemoteStream;
+import io.camunda.zeebe.transport.stream.impl.ImmutableStreamRegistry.StreamConsumer;
 import io.camunda.zeebe.util.buffer.BufferReader;
 import io.camunda.zeebe.util.buffer.BufferWriter;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ThreadLocalRandom;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public class RemoteStreamImpl<M extends BufferReader, P extends BufferWriter>
     implements RemoteStream<M, P> {
 
-  private final M metadata;
+  private static final Logger LOGGER = LoggerFactory.getLogger(RemoteStreamImpl.class);
+  private final AggregatedRemoteStream<M> stream;
   private final RemoteStreamPusher<P> streamer;
+  private final Executor executor;
 
-  public RemoteStreamImpl(final M metadata, final RemoteStreamPusher<P> streamer) {
-    this.metadata = metadata;
+  public RemoteStreamImpl(
+      final AggregatedRemoteStream<M> stream,
+      final RemoteStreamPusher<P> streamer,
+      final Executor executor) {
+    this.stream = stream;
     this.streamer = streamer;
+    this.executor = executor;
   }
 
   @Override
   public M metadata() {
-    return metadata;
+    return stream.getLogicalId().metadata();
   }
 
   @Override
   public void push(final P payload, final ErrorHandler<P> errorHandler) {
-    streamer.pushAsync(payload, errorHandler);
+    executor.execute(
```

Comment:
I checked your code. One main thing I was considering is that the common case is without any retry. First attempt should succeed. So generating a random order by `Collections::shuffle` like an overhead, because in most cases we won't have to use it. This might be a pre-mature optimization, though.

Refreshing set of stream consumers is useful still.

Reply:
That's a good point. We could always start with a random one (so just generating a random integer) and add it immediately to the attempted, and shuffle only on retry (though it may be awkward to shuffle all the time - but as you said, the idea is we don't need to retry usually).

Reply:
:+1: Let's improve this in a follow up.